File: Smart_Park-main/parking pal/parking_system.py
from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from typing import Dict, List, Optional
import uuid

class VehicleType(Enum):
    MOTORCYCLE = 1
    CAR = 2
    SUV = 3
    TRUCK = 4

@dataclass
class Vehicle:
    license_plate: str
    vehicle_type: VehicleType
    owner_id: str
    entry_time: datetime

class SmartParkingSystem:
    def __init__(self):
        self.spots: Dict[int, bool] = {}  # spot_id -> occupied
        self.vehicle_map: Dict[str, int] = {}  # license -> spot_id
        self.waiting_list: List[Vehicle] = []
        self.spot_sensors: Dict[int, bool] = {}  # spot_id -> sensor_status
        self.revenue: float = 0.0

    def initialize_spots(self, num_spots: int):
        """Initialize parking spots with IoT sensors"""
        for i in range(num_spots):
            self.spots[i] = False
            self.spot_sensors[i] = True  # Sensor active

    def find_nearest_spot(self, vehicle_type: VehicleType) -> Optional[int]:
        """Find the nearest available spot using sensor data"""
        for spot_id, occupied in self.spots.items():
            if not occupied and self.spot_sensors[spot_id]:
                if self._is_spot_suitable(spot_id, vehicle_type):
                    return spot_id
        return None

    def park_vehicle(self, vehicle: Vehicle) -> dict:
        """Park a vehicle with smart allocation"""
        spot_id = self.find_nearest_spot(vehicle.vehicle_type)
        
        if spot_id is None:
            self.waiting_list.append(vehicle)
            return {
                "status": "waiting",
                "position": len(self.waiting_list),
                "estimated_wait": self._estimate_wait_time()
            }

        self.spots[spot_id] = True
        self.vehicle_map[vehicle.license_plate] = spot_id

        return {
            "status": "parked",
            "spot_id": spot_id,
            "entry_time": datetime.now(),
            "ticket_id": str(uuid.uuid4())
        }

    def exit_vehicle(self, license_plate: str) -> dict:
        """Process vehicle exit with automatic payment"""
        if license_plate not in self.vehicle_map:
            raise ValueError("Vehicle not found")

        spot_id = self.vehicle_map[license_plate]
        self.spots[spot_id] = False
        del self.vehicle_map[license_plate]

        # Process waiting list
        self._process_waiting_list()

        return {
            "status": "success",
            "exit_time": datetime.now(),
            "fee": self._calculate_fee(license_plate)
        }

    def _process_waiting_list(self):
        """Process vehicles in waiting list"""
        if not self.waiting_list:
            return

        vehicle = self.waiting_list[0]
        spot_id = self.find_nearest_spot(vehicle.vehicle_type)
        
        if spot_id is not None:
            self.waiting_list.pop(0)
            self.park_vehicle(vehicle) 

File: Smart_Park-main/parking pal/parking_system.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>

#define MAX_SPOTS 1000
#define MAX_VEHICLES 2000
#define PLATE_LENGTH 10

typedef enum {
    MOTORCYCLE = 0,
    CAR,
    SUV,
    TRUCK
} VehicleType;

typedef struct {
    int id;
    VehicleType type;
    int floor;
    int is_occupied;
    time_t occupied_since;
} ParkingSpot;

typedef struct {
    char license_plate[PLATE_LENGTH];
    VehicleType type;
    int spot_id;
    time_t entry_time;
} Vehicle;

typedef struct {
    ParkingSpot* spots[MAX_SPOTS];
    Vehicle* vehicles[MAX_VEHICLES];
    int spot_count;
    int vehicle_count;
    double revenue;
} ParkingSystem;

ParkingSystem* initialize_parking_system(int num_spots) {
    ParkingSystem* system = (ParkingSystem*)malloc(sizeof(ParkingSystem));
    system->spot_count = num_spots;
    system->vehicle_count = 0;
    system->revenue = 0.0;

    for (int i = 0; i < num_spots; i++) {
        system->spots[i] = (ParkingSpot*)malloc(sizeof(ParkingSpot));
        system->spots[i]->id = i;
        system->spots[i]->is_occupied = 0;
        system->spots[i]->floor = i / 100; // 100 spots per floor
        system->spots[i]->type = i % 4; // Distribute spot types
    }

    return system;
}

int find_optimal_spot(ParkingSystem* system, VehicleType type) {
    // Use bit manipulation for faster processing
    int spot_mask = 1 << type;
    
    for (int i = 0; i < system->spot_count; i++) {
        if (!system->spots[i]->is_occupied && 
            (spot_mask & (1 << system->spots[i]->type))) {
            return i;
        }
    }
    return -1;
}

double calculate_fee(time_t entry_time, VehicleType type) {
    time_t exit_time = time(NULL);
    double duration = difftime(exit_time, entry_time) / 3600.0; // Hours
    double base_rate = 2.0;
    
    switch (type) {
        case MOTORCYCLE:
            return duration * base_rate;
        case CAR:
            return duration * base_rate * 1.5;
        case SUV:
            return duration * base_rate * 2.0;
        case TRUCK:
            return duration * base_rate * 3.0;
        default:
            return 0.0;
    }
}

int park_vehicle(ParkingSystem* system, const char* license_plate, VehicleType type) {
    if (system->vehicle_count >= MAX_VEHICLES) {
        return -1; // System full
    }

    int spot_id = find_optimal_spot(system, type);
    if (spot_id == -1) {
        return -2; // No suitable spot
    }

    Vehicle* vehicle = (Vehicle*)malloc(sizeof(Vehicle));
    strncpy(vehicle->license_plate, license_plate, PLATE_LENGTH - 1);
    vehicle->type = type;
    vehicle->spot_id = spot_id;
    vehicle->entry_time = time(NULL);

    system->vehicles[system->vehicle_count++] = vehicle;
    system->spots[spot_id]->is_occupied = 1;
    system->spots[spot_id]->occupied_since = vehicle->entry_time;

    return spot_id;
} 